### ISSUE TITLE
Navigate between files using namespace names

### ISSUE THREAD

**Issue by markaddleman** [BetterThanTomorrow/calva#545]
In Cursive, you can navigate among files by using their namespace names rather than munged directory names.  For example, I could navigate to `library.test-utils` instead of `library/test_utils`.  It's a small thing but it helps to reduce context switching

**Comment by bpringe** [comment]
This could probably be implemented as a custom command that prompts for a namespace, finds the associated file in the project directory, and opens it in an editor.

**Comment by bpringe** [comment]
@markaddleman Is this done statically in Cursive (can you navigate like that without a repl connection)?

**Comment by markaddleman** [comment]
Yes

**Comment by bpringe** [comment]
@ericdallo, can and should this be implemented in clojure-lsp?

**Comment by PEZ** [comment]
I'm thinking on when this would actually be needed much:

1. When typing in the file browser you can just skip typing the dots and dashes and you will find the file pretty easily.
2. If it is a namespace in a Clojure file, you can navigate to it (e.g. cmd/ctrl-click).
3. If it is that you have it some other file and want to paste it, well, I just made an extension that can help with that, and added an example for this use case: https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.paste-replaced

Is there some base not covered by these? A demo gif of how its done in Cursive would be nice!

**Comment by ericdallo** [comment]
> A demo gif of how its done in Cursive would be nice!

Yeah, that would certainly help, I can't easily see how that would work

**Comment by bpringe** [comment]
@markaddleman Is this what you were referring to? (Taken from [here](https://cursive-ide.com/userguide/navigation.html#navigate-to-files-symbols-and-namespaces).)

![image](https://user-images.githubusercontent.com/12722744/133509202-5da515f6-8919-46f3-bfb1-5a9810580e93.png)

Regarding what @PEZ said, it seems we are not lacking in options for navigation. In any case though, I think some clarification would be helpful on what is being requested exactly.

**Comment by ericdallo** [comment]
@bpringe if so, this already works via the `workspace/symbol` request, this is an example using lsp-mode `lsp-ivy-workspace-symbol`

![2021-09-15_18-19](https://user-images.githubusercontent.com/7820865/133510986-d75c47cc-556c-4da8-aa19-2592714b174e.png)

**Comment by markaddleman** [comment]
> @markaddleman Is this what you were referring to? (Taken from [here](https://cursive-ide.com/userguide/navigation.html#navigate-to-files-symbols-and-namespaces).)
> 
> ![image](https://user-images.githubusercontent.com/12722744/133509202-5da515f6-8919-46f3-bfb1-5a9810580e93.png)
> 
> Regarding what @PEZ said, it seems we are not lacking in options for navigation. In any case though, I think some clarification would be helpful on what is being requested exactly.

Yes, this is what I was thinking.  Navigating namespaces by clicking through references in source files is also ab important feature for me

**Comment by ericdallo** [comment]
@bpringe that is already possible to be implemented as I showed before, let me know if have any questions about how lsp-mode works for that.

> Navigating namespaces by clicking through references in source files is also ab important feature for me

This is a LSP protocol limitation, the lenses click/action only allows to execute a `Command`, not offer a feedback to user showing the references or something like that, I need to investigate that further more to see if there any other choices/workarounds

**Comment by mrkam2** [comment]
I'm adding my voice to this feature. The scenario I need this the most often is when going from test execution logs and stacktraces to the code. Unfortunately, these often feature namespace names rather than complete file paths. 

Stacktrace example:
```
              compojure.core$wrap_response$fn__18493.invoke (core.clj:158)
```
This line only specifies the file name which is quite generic, so you need to use the namespace name to jump to the correct file.

Test log example:
```
Testing reporter.reports-test
```
This one only has the namespace.

@PEZ implemented a really cool [workaround](https://marketplace.visualstudio.com/items?itemName=betterthantomorrow.paste-replaced) for this, which is altering paste operation to replace certain characters. I was able to configure it to work really nicely with one caveat that it still requires you to double-check the match as it could choose a different one because of the recency or other artifacts of search results sorting. See the discussion in [slack](https://clojurians.slack.com/archives/CBE668G4R/p1631718868200800).

**Comment by ericdallo** [comment]
@bpringe I just added the support for clicking on references code lens for emacs/lsp-mode [here](https://github.com/emacs-lsp/lsp-mode/pull/3106), Unfortunately, it's something that must be done for every client, handle that command and show call the references from the command args.
I suggest you create a new issue on Calva to add that support

**Comment by bpringe** [comment]
@ericdallo I may be misunderstanding what you mean, but I think Calva has this feature already. We can click the references code lens and see the references and click them.

![image](https://user-images.githubusercontent.com/12722744/133870418-9acdf4d0-fe6d-41e1-bb54-bbe729c2fd16.png)

**Comment by bpringe** [comment]
Regarding the original post, it looks like we can implement this how Eric said [here](https://github.com/BetterThanTomorrow/calva/issues/545#issuecomment-920391275). I'm sort of just making this post for added visibility later.

**Comment by bpringe** [comment]
If anyone wants to submit a PR for this, I'd be willing to show you around the code and how to work with clojure-lsp from Calva.

@ericdallo, in [that image](https://github.com/BetterThanTomorrow/calva/issues/545#issuecomment-920391275), is that showing in lsp-mode you can start typing a workspace symbol and it will call `workspace/symbol` and display those results for the user to choose from?

**Comment by ericdallo** [comment]
> @ericdallo, in that image, is that showing in lsp-mode you can start typing a workspace symbol and it will call workspace/symbol and display those results for the user to choose from?

Yes, clojure-lsp has a fuzzy search in the backend, and each keystroke lsp-mode search request the workspace/symbol with the custom query as per the SPEC.

![lsp-ivy-workspace-symbol](https://user-images.githubusercontent.com/7820865/133890099-e5fa2e0d-9996-4006-9c46-3a0b88d7dcde.gif)

**Comment by ericdallo** [comment]
> @ericdallo I may be misunderstanding what you mean, but I think Calva has this feature already. We can click the references code lens and see the references and click them.

Oh i see, that was already implemented, very clever :)
https://github.com/BetterThanTomorrow/calva/blob/published/src/lsp/main.ts#L195-L198

**Comment by bpringe** [comment]
Thanks 😄
